app/uk/gov/hmrc/economiccrimelevyregistration/controllers/RegistrationSubmissionController.scala

```diff
@@ -16,129 +16,118 @@
-                  additionalInfo.eclReference match {
-                    case Some(eclRef) =>
-                      dmsService
-                        .submitToDms(registration.base64EncodedFields.flatMap(_.dmsSubmissionHtml), Instant.now())
-                        .flatMap {
-                          case Right(response) =>
-                            if (appConfig.amendRegistrationNrsEnabled) {
-                              nrsService.submitToNrs(
-                                registration.base64EncodedFields.flatMap(_.nrsSubmissionHtml),
-                                eclRef,
-                                appConfig.eclAmendRegistrationNotableEvent
-                              )
-                            }
-                            auditService
-                              .successfulSubscriptionAndEnrolment(
-                                registration,
-                                eclRef,
-                                additionalInfo.liabilityYear
-                              )
-                            Future.successful(Ok(Json.toJson(response)))
-                          case Left(e)         =>
-                            logger.error(
-                              s"Failed to submit PDF to DMS: ${e.getMessage()}"
-                            )
-                            Future.successful(InternalServerError("Could not send PDF to DMS queue"))
+    (for {
+      registration   <- registrationService.getRegistration(id).asResponseError
+      additionalInfo <- registrationAdditionalInfoService.get(registration.internalId).asResponseError
+      _              <- (if (registration.isRegistration) {
+                           registrationValidation(registration)
+                         } else {
+                           subscriptionValidation(registration, additionalInfo)
+                         }).asResponseError
+      response       <- if (registration.isRegistration) {
+                          registerForEcl(registration, additionalInfo.liabilityYear)
+                        } else {
+                          subscribeToEcl(registration, additionalInfo)
                         }
-                    case None         =>
-                      logger.error(
-                        s"Expected eclReference to be present in additional information for amendment with internal id: ${registration.internalId}"
-                      )
+    } yield response).convertToResult(OK)
+  }
 
-                      Future.successful(
-                        InternalServerError("Expected eclReference to be present in additional information")
-                      )
+  private def registrationValidation(registration: Registration): EitherT[Future, DataValidationError, Registration] =
+    EitherT {
+      Future.successful(
+        registrationValidationService.validateRegistration(registration) match {
```

Comment:
Why this `match` as you are returning what you are matching against?

Reply:
Refactored :)